src/contract.ts

```
@@ -0,0 +1,381 @@
+      Buffer.from('00000000', 'hex'),
+      encodeSeeds(seeds),
+      Buffer.from(input.replace('0x', ''), 'hex'),
+    ]);
+
+    const keys = [
+      ...seeds.map((seed) => ({
+        pubkey: seed.address,
+        isSigner: false,
+        isWritable: true,
+      })),
+      {
+        pubkey: this.contractStorageAccount.publicKey,
+        isSigner: false,
+        isWritable: true,
+      },
+      {
+        pubkey: SYSVAR_CLOCK_PUBKEY,
+        isSigner: false,
+        isWritable: false,
+      },
+      {
+        pubkey: PublicKey.default,
+        isSigner: false,
+        isWritable: false,
+      },
+      ...accounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: false,
+      })),
+      ...writableAccounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: true,
+      })),
+    ];
+
+    const instruction = new TransactionInstruction({
+      keys,
+      programId: this.program.programAccount.publicKey,
+      data,
+    });
+
+    signers.unshift(this.program.payerAccount);
+
+    const { encoded, logs, computeUnitsUsed } = await this.program.makeTx(
+      simulate,
+      instruction,
+      signers
+    );
+
+    let result: ethers.utils.Result | null = null;
+
+    if (fragment.outputs?.length) {
+      if (!encoded) {
+        throw new Error('return data not set');
+      }
+
+      result = this.abi.decodeFunctionResult(fragment, encoded);
+      if (fragment.outputs.length === 1) {
+        result = result[0];
+      }
+    }
+
+    const events = this.parseLogsEvents(logs);
+
+    return { result, logs, computeUnitsUsed, events };
+  }
+
+  /**
+   *
+   * @param test
+   * @param upto
+   * @returns
+   */
+  async contractStorage(test: Program, upto: number): Promise<Buffer> {
+    const accountInfo = await test.connection.getAccountInfo(
+      this.contractStorageAccount.publicKey
+    );
+
+    return accountInfo!.data;
```

If `accountInfo` can be nullish, it seems better to throw a known error.